python/cmsis_svd/model.py

```diff
@@ -1,452 +1,620 @@
-            value=value,
-            expected_type=expected_type,
-            actual_type=type(value),
-        ))
-    return value
+"""
+
+from typing import List, Optional, Any, Union, TypedDict, Generator
+from dataclasses import dataclass, field, fields
+from enum import Enum
+import inspect
+import copy
+import json
+import six
 
 
-def _none_as_empty(v):
+################################################################################
+# Utils
+################################################################################
+
+def _none_as_empty(v: Any) -> Generator[Any, None, None]:
     if v is not None:
         for e in v:
             yield e
 
 
+def isinstance_by_str(instance: Any, type_name: str) -> bool:
+    if (cls := globals().get(type_name, None)) and inspect.isclass(cls):
+        return isinstance(instance, cls)
+    return False
+
+
+################################################################################
+# CMSIS-SVD Model Base class
+################################################################################
+
 class SVDJSONEncoder(json.JSONEncoder):
-    _TO_DICT_SKIP_KEYS = {"register_arrays", "parent"}
+    _TO_DICT_SKIP_KEYS = {'parent'}
 
-    def default(self, obj):
-        if isinstance(obj, SVDElement):
+    def default(self, o: Any) -> Any:
+        if isinstance(o, SVDElement):
             eldict = {}
-            for k, v in six.iteritems(obj.__dict__):
+            for k, v in six.iteritems(o.__dict__):
                 if k in self._TO_DICT_SKIP_KEYS:
                     continue
                 if k.startswith("_"):
                     pubkey = k[1:]
-                    eldict[pubkey] = getattr(obj, pubkey)
+                    eldict[pubkey] = getattr(o, pubkey)
                 else:
                     eldict[k] = v
             return eldict
-        else:
-            return json.JSONEncoder.default(self, obj)
 
+        if isinstance(o, Enum):
+            return o.value
 
-class SVDElement(object):
-    """Base class for all SVD Elements"""
+        return json.JSONEncoder.default(self, o)
 
-    def __init__(self):
-        self.parent = None
 
-    def to_dict(self):
+@dataclass
+class SVDElement:
+    """Base class for all SVD Elements"""
+    parent: Optional['SVDElement'] = field(default=None)
+
+    def to_dict(self) -> Any:
         # This is a little convoluted but it works and ensures a
         # json-compatible dictionary representation (at the cost of
         # some computational overhead)
         encoder = SVDJSONEncoder()
         return json.loads(encoder.encode(self))
 
 
+################################################################################
+# CMSIS-SVD Model Type
+################################################################################
+
+class SVDAccessType(Enum):
+    READ_ONLY = 'read-only'
+    WRITE_ONLY = 'write-only'
+    READ_WRITE = 'read-write'
+    WRITE_ONCE = 'writeOnce'
```

Comment:
Consistency with other attributes of this enum.

```suggestion
    WRITE_ONCE = 'write-once'
```

Reply:
see [registerPropertiesGroup documention](https://open-cmsis-pack.github.io/svd-spec/main/elem_special.html), the cmsis specification uses inconsistent names :/ .